tsdb/db_test.go

```diff
@@ -6334,3 +6352,54 @@ func compareSeries(t require.TestingT, expected, actual map[string][]tsdbutil.Sa
 		}
 	}
 }
+
+func TestChunkQuerierReadWriteRace(t *testing.T) {
```

Comment:
If you run this test with removed re-encoding in qurier.go, but without the chunk copying in head_read.go, then this test will show you the race. Copying the chunk takes away the race.

Reply:
(Goes without saying that this test has no race without any changes, i.e. with re-encoding intact)